src/main/drivers/temperature/ds18b20.c

```diff
@@ -0,0 +1,141 @@
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#include "build/debug.h"
+
+#include "drivers/1-wire.h"
+#include "drivers/1-wire/ds2482.h"
+#include "drivers/1-wire/ds_crc.h"
+#include "drivers/temperature/ds18b20.h"
+
+
+#define DS18B20_FAMILY_CODE 0x28
+
+#define DS18B20_ALARM_SEARCH_CMD 0xEC
+#define DS18B20_START_CONVERSION_CMD 0x44
+#define DS18B20_WRITE_SCRATCHPAD_CMD 0x4E
+#define DS18B20_READ_SCRATCHPAD_CMD 0xBE
+#define DS18B20_COPY_SCRATCHPAD_CMD 0x48
+#define DS18B20_RECALL_EEPROM_CMD 0xB8
+#define DS18B20_READ_POWER_SUPPLY_CMD 0xB4
+
+#if defined(USE_1WIRE) && defined(USE_1WIRE_DS2482) && defined(USE_TEMPERATURE_DS18B20)
+
+#define _1WireDev (&ds2482Dev)
+
+bool ds18b20_enumerate(uint64_t *rom_table, uint8_t *rom_table_len)
+{
+    return ds2482_1wire_search_rom(_1WireDev, DS18B20_FAMILY_CODE, rom_table, rom_table_len);
+}
+
+bool ds18b20_configure(uint64_t rom, uint8_t config)
+{
+    bool ack = ds2482_1wire_reset_and_match_rom(_1WireDev, rom);
+    if (!ack) return false;
+
+    uint8_t buf[4] = { DS18B20_WRITE_SCRATCHPAD_CMD, 0, 0, config };
+    return ds2482_1wire_write_buf(_1WireDev, buf, sizeof(buf));
+}
+
+static bool ds18b20_parasitic_powered_present()
+{
+    bool ack = ds2482_1wire_reset_and_skip_rom(_1WireDev);
+    if (!ack) return false;
+
+    ack = ds2482_1wire_write_byte(_1WireDev, DS18B20_READ_POWER_SUPPLY_CMD);
+    if (!ack) return false;
+
+    uint8_t status;
+    ack = ds2482_poll(_1WireDev, true, &status);
+    if (!ack) return false;
+
+    return !DS2482_SBR_VALUE(status);
+}
+
+static bool ds18b20_read_scratchpad()
+{
+    return ds2482_1wire_write_byte(_1WireDev, DS18B20_READ_SCRATCHPAD_CMD);
+}
+
+static bool ds18b20_read_scratchpad_buf(uint8_t *buf, uint8_t len)
+{
+    ds18b20_read_scratchpad(_1WireDev);
+    ds2482_wait_for_bus(_1WireDev);
+
+    for (uint8_t index = 0; index < len; ++index) {
+        bool ack = ds2482_1wire_read_byte(_1WireDev);
+        if (!ack) return false;
+
+        ack = ds2482_wait_for_bus(_1WireDev);
+        if (!ack) return false;
+
+        ack = ds2482_read_byte(_1WireDev, buf + index);
+        if (!ack) return false;
+    }
+    return true;
+}
+
+bool ds18b20_start_conversion()
+{
+    bool parasitic_power = ds18b20_parasitic_powered_present(_1WireDev);
+    if (parasitic_power) return false;
+    bool ack = ds2482_1wire_reset_and_skip_rom(_1WireDev);
+    if (!ack) return false;
+    return ds2482_1wire_write_byte(_1WireDev, DS18B20_START_CONVERSION_CMD);
+}
+
+bool ds18b20_wait_for_conversion()
```

Comment:
Yes I need to take a look into it because even though everything works fine on the bench after checking the protocol timing reading the sensor values takes too much time. The reading process will need to yield before all the sensors are read or even during addressing/reading. 1-Wire is pretty slow. Only addressing the sensor takes 7.7ms. Reading the temperature value then takes 8.1ms. So we are talking a bit less than 16ms per sensor to read back the temperature value. Fortunately we can send the "start conversion" to all the sensors at once and that only takes ~2.2ms.

Reply:
Resetting the 1-Wire bus (needed before every command) is taking 1345µs in itself so it looks like we will need to. We need to send the reset command (54µs) then yield and the next iterations check if the bus has been reset (54µs) then send a new command